random: Clarify the semantics of TEST_RANDOM_GENERATOR

stephanosio opened the pull request:
The `TEST_RANDOM_GENERATOR` Kconfig symbol indicates that a non-random
number generator (i.e. a random number generator that is not truly
random) is _allowed_ to be used for testing purposes, and not
necessarily that a non-random number generator is _selected_ -- that is
indicated by the `RNG_GENERATOR_CHOICE` choices that depend on
`TEST_RANDOM_GENERATOR` (i.e. only `TIMER_RANDOM_GENERATOR` as of now).

This commit updates the `TEST_RANDOM_GENERATOR` Kconfig symbol
description to reflect and clarify that.

It also removes the `TEST_RANDOM_GENERATOR`'s dependency on
`ENTROPY_HAS_DRIVER` because the act of _allowing_ a non-random number
generator to be used does not depend on the availability of an entropy
driver -- when an entropy driver is available, by default, the
`ENTROPY_DEVICE_RANDOM_GENERATOR` will be selected; otherwise,
`TIMER_RANDOM_GENERATOR`, a non-random generator, will be selected.

---

Closes #53019

stephanosio committed:
random: Clarify the semantics of TEST_RANDOM_GENERATOR

The `TEST_RANDOM_GENERATOR` Kconfig symbol indicates that a non-random
number generator (i.e. a random number generator that is not truly
random) is _allowed_ to be used for testing purposes, and not
necessarily that a non-random number generator is _selected_ -- that is
indicated by the `RNG_GENERATOR_CHOICE` choices that depend on
`TEST_RANDOM_GENERATOR` (i.e. only `TIMER_RANDOM_GENERATOR` as of now).

This commit updates the `TEST_RANDOM_GENERATOR` Kconfig symbol
description to reflect and clarify that.

It also removes the `TEST_RANDOM_GENERATOR`'s dependency on
`ENTROPY_HAS_DRIVER` because the act of _allowing_ a non-random number
generator to be used does not depend on the availability of an entropy
driver -- when an entropy driver is available, by default, the
`ENTROPY_DEVICE_RANDOM_GENERATOR` will be selected; otherwise,
`TIMER_RANDOM_GENERATOR`, a non-random generator, will be selected.

Signed-off-by: Stephanos Ioannidis <stephanos.ioannidis@nordicsemi.no>

stephanosio committed:
random: Fix non-random number generator warning condition

The `TEST_RANDOM_GENERATOR` symbol in itself does not imply that a
non-random number generator is selected -- that is indicated by the
`RNG_GENERATOR_CHOICE` choices, of which only `TIMER_RANDOM_GENERATOR`
is a non-random generator.

This commit updates the Random subsystem to print the "non-random
number generator usage warning" when `TIMER_RANDOM_GENERATOR` is
selected, as opposed to when `TEST_RANDOM_GENERATOR` is selected.

Signed-off-by: Stephanos Ioannidis <stephanos.ioannidis@nordicsemi.no>

frkv commented on subsys/random/Kconfig:
```diff
@@ -6,18 +6,32 @@
 	prompt "Random generator"
-	default ENTROPY_DEVICE_RANDOM_GENERATOR
+	default ENTROPY_DEVICE_RANDOM_GENERATOR if ENTROPY_HAS_DRIVER
+	default TIMER_RANDOM_GENERATOR if TEST_RANDOM_GENERATOR
```
The fallback to utilize a timer peripheral as a "pseudo entropy source" has security implications in your system. There exists as-of-now very little protection against this ending up in an end-product due to misconfigured DTS or driver configuration. 

I have argued in a related feature request that there is an option available to us that goes beyond just printing warnings once TEST_RANDOM_ENTROPY/TIMER_RANDOM_GENERATOR is enabled in the system 

Please see: https://github.com/zephyrproject-rtos/zephyr/issues/53019#issuecomment-1350956828

stephanosio replied:
> There exists as-of-now very little protection against this ending up in an end-product due to misconfigured DTS or driver configuration.

IMHO, a very intrusive warning that is printed every time you run CMake is quite sufficient.

> I have argued in a related feature request that there is an option available to us that goes beyond just printing warnings once TEST_RANDOM_ENTROPY/TIMER_RANDOM_GENERATOR is enabled in the system

That is well beyond the scope of the enhancement issue or this PR. That will require a much broader discussion on how to organise and handle different build profiles (preferably, with a detaild RFC on what that would look like).

frkv replied:
There is a reference to `hardenconfig` which already exists that provides exactly what I wish for (see PR comment) https://github.com/zephyrproject-rtos/zephyr/issues/53019#issuecomment-1350956828) which is showing that TEST_RANDOM_GENERATOR is disallowed when you build with hardenconfig. I would argue that an additional line should be made in this file to also disqualify TIMER_RANDOM_GENERATOR. This could be done in this PR...

https://github.com/zephyrproject-rtos/zephyr/blob/95717520bf4309c7e54fdf6d99975491880fb874/scripts/kconfig/hardened.csv#LL27-L27C22

Please add `TIMER_RANDOM_GENERATOR,n` then my comment on fallback for RNG_GENERATOR_CHOICE is sufficiently addressed

stephanosio replied:
> I propose adding `TIMER_RANDOM_GENERATOR,n` to the hardenconfig.

Done.

carlescufi commented:
@ceolin @joerchan could you please take a look?

joerchan reviewed (dismissed):
I think we should remove the instances in the zephyr tree where the TEST_RANDOM_GENERATOR is enabled in addition to this change.

The bluetooth IPSP sample is an example of this being enabled where it should not have been.

stephanosio commented:
> The bluetooth IPSP sample is an example of this being enabled where it should not have been.

@joerchan If you are referring to `samples/bluetooth/ipsp`, I am not sure if I understand the rationale. [That sample runs on `qemu_x86` and `qemu_cortex_m3`](https://github.com/zephyrproject-rtos/zephyr/blob/4b2fffc7f98e735d99eea790342113fd82fcd66b/samples/bluetooth/ipsp/sample.yaml#L5-L7) and therefore requires `CONFIG_TEST_RANDOM_GENERATOR=y`.

In general, tests and samples should set `CONFIG_TEST_RANDOM_GENERATOR=y` so that they can run on a wide range of devices (i.e. on the devices without an entropy device) -- that is what `CONFIG_TEST_RANDOM_GENERATOR` is for after all. Even if `CONFIG_TEST_RANDOM_GENERATOR=y`, the test/sample will run with a true RNG if an entropy device is available.

The NCS samples are rather a special case in the sense that they only need to run on the nRF devices, which comes with an entropy device, and not any other devices, in which case you can get away with `CONFIG_TEST_RANDOM_GENERATOR=n`.

joerchan commented:
> > The bluetooth IPSP sample is an example of this being enabled where it should not have been.
> 
> @joerchan If you are referring to `samples/bluetooth/ipsp`, I am not sure if I understand the rationale. [That sample runs on `qemu_x86` and `qemu_cortex_m3`](https://github.com/zephyrproject-rtos/zephyr/blob/4b2fffc7f98e735d99eea790342113fd82fcd66b/samples/bluetooth/ipsp/sample.yaml#L5-L7) and therefore requires `CONFIG_TEST_RANDOM_GENERATOR=y`.
> 
> In general, tests and samples should set `CONFIG_TEST_RANDOM_GENERATOR=y` so that they can run on a wide range of devices (i.e. on the devices without an entropy device) -- that is what `CONFIG_TEST_RANDOM_GENERATOR` is for after all. Even if `CONFIG_TEST_RANDOM_GENERATOR=y`, the test/sample will run with a true RNG if an entropy device is available.
> 
> The NCS samples are rather a special case in the sense that they only need to run on the nRF devices, which come with an entropy device, and not any other devices, in which case you can get away with `CONFIG_TEST_RANDOM_GENERATOR=n`.

This is a bluetooth sample. It should be able to use the entropy of the bluetooth controller `ENTROPY_BT_HCI`.

stephanosio committed:
hardening: Require CONFIG_TIMER_RANDOM_GENERATOR=n

This commit adds the requirement for the `TIMER_RANDOM_GENERATOR`
Kconfig to be set to `n` because it enables timer-based pseudo-random
number generator, which is only intended to be used for testing
purposes.

Signed-off-by: Stephanos Ioannidis <stephanos.ioannidis@nordicsemi.no>

joerchan commented:
> > The sample itself does not have any entropy requirements, so not sure it needs it.
> 
> `sys_rand_get` requires a RNG and that function is used at least once in that sample. For instance, building with `CONFIG_TEST_RANDOM_GENERATOR=n` results in build failures for `qemu_x86`:
> 
> ```
> /opt/zephyr-sdk-0.15.2/x86_64-zephyr-elf/bin/../lib/gcc/x86_64-zephyr-elf/12.1.0/../../../../x86_64-zephyr-elf/bin/ld.bfd: zephyr/subsys/net/ip/libsubsys__net__ip.a(net_if.c.obj): in function `sys_rand32_get':
> /home/stephanos/Dev/zephyrproject/zephyr/twister-out/qemu_x86/zephyr/samples/bluetooth/ipsp/sample.bluetooth.ipsp/zephyr/include/generated/syscalls/rand32.h:32: undefined reference to `z_impl_sys_rand32_get'
> ```
> 
> I will leave the sample as is for now.

Ok. It could still work with the BT entropy source though.
```
cp ./tests/drivers/entropy/api/entropy_bt_hci.overlay ./samples/bluetooth/ipsp/entropy_bt_hci.overlay 
west build -b qemu_x86 samples/bluetooth/ipsp -- -DCONFIG_TEST_RANDOM_GENERATOR=n -DDTC_OVERLAY_FILE=./entropy_bt_hci.overlay
```

Leaving it as is is OK with me though. Adding overlay files is a very complicated way to get an entropy source in your application.

ceolin reviewed (approved):
LGTM
